flink-examples/flink-examples-batch/src/main/java/org/apache/flink/examples/java/ap/AffinityPropagationBulk.java

```diff
@@ -0,0 +1,449 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+package org.apache.flink.examples.java.ap;
+
+import org.apache.flink.api.common.aggregators.ConvergenceCriterion;
+import org.apache.flink.api.common.aggregators.LongSumAggregator;
+import org.apache.flink.api.common.functions.FilterFunction;
+import org.apache.flink.api.common.functions.GroupReduceFunction;
+import org.apache.flink.api.common.functions.JoinFunction;
+import org.apache.flink.api.common.functions.MapFunction;
+import org.apache.flink.api.common.functions.RichJoinFunction;
+import org.apache.flink.api.common.operators.Order;
+import org.apache.flink.api.java.DataSet;
+import org.apache.flink.api.java.ExecutionEnvironment;
+import org.apache.flink.api.java.operators.IterativeDataSet;
+import org.apache.flink.api.java.tuple.Tuple3;
+import org.apache.flink.api.java.tuple.Tuple4;
+import org.apache.flink.examples.java.ap.util.AffinityPropagationData;
+import org.apache.flink.types.DoubleValue;
+import org.apache.flink.types.LongValue;
+import org.apache.flink.util.Collector;
+import org.apache.flink.api.java.functions.FunctionAnnotation.ForwardedFieldsFirst;
+import org.apache.flink.api.java.functions.FunctionAnnotation.ForwardedFieldsSecond;
+import org.apache.flink.api.java.functions.FunctionAnnotation.ForwardedFields;
+
+/**
+ * Created by joseprubio on 9/22/16.
+ */
+
+public class AffinityPropagationBulk {
+
+	private static final double DAMPING_FACTOR = 0.9;
+	private static final double CONVERGENCE_THRESHOLD = 0.12;
+	private static final String CONVERGENCE_MESSAGES = "message convergence";
+
+	public static void main(String[] args) throws Exception {
+
+		ExecutionEnvironment env = ExecutionEnvironment.getExecutionEnvironment();
+		env.getConfig().enableObjectReuse();
+
+		// Get input similarities Tuple3<src, target, similarity>
+		DataSet<Tuple3<LongValue, LongValue, DoubleValue>> similarities =
+			AffinityPropagationData.getTuplesFromFile(env);
+
+		// Init input to iteration
+		DataSet<Tuple4<LongValue, LongValue, DoubleValue, DoubleValue>> initMessages
+			= similarities.map(new InitMessage());
+
+		// Iterate
+		IterativeDataSet<Tuple4<LongValue, LongValue, DoubleValue, DoubleValue>> messages
+			= initMessages.iterate(20);
+
+		// Create aggregator
+		messages.registerAggregationConvergenceCriterion(CONVERGENCE_MESSAGES, new LongSumAggregator(),
+			new MessageConvergence(similarities.count() * 2));
+
+		// Start responsibility message calculation
+		// r(i,k) <- s(i,k) - max {a(i,K) + s(i,K)} st K != k
+		// Iterate over Tuple6 <Source, Target, Responsibility , Availability, IsExemplar, ConvergenceCounter>
```

Comment:
Would "exemplar" be better than "target" and maybe "vertex" rather than "source"? Why is `ConvergenceCounter` here?